File: sdk/typescript/test/env.test.ts
import { describe, expect, it } from 'vitest';

import { normalizePortName, readEnv } from '../src/env';

describe('normalizePortName', () => {
  it('uppercases and replaces spaces with underscores', () => {
    expect(normalizePortName('voice chat')).toBe('VOICE_CHAT');
    expect(normalizePortName('  query ')).toBe('QUERY');
    expect(normalizePortName('Voice')).toBe('VOICE');
  });
});

describe('readEnv', () => {
  it('reads the default port', () => {
    const env = readEnv({ GAMEFLOW_DEFAULT_PORT: '7777' });
    expect(env.ports.default).toBe(7777);
  });

  it('reads named ports with normalization', () => {
    const env = readEnv({ GAMEFLOW_VOICE_CHAT_PORT: '9000' });
    expect(env.ports.get('voice chat')).toBe(9000);
    expect(env.ports.get('VOICE_CHAT')).toBe(9000);
  });

  it('reads TLS ports separately', () => {
    const env = readEnv({
      GAMEFLOW_DEFAULT_PORT: '7777',
      GAMEFLOW_TLS_DEFAULT_PORT: '8443',
      GAMEFLOW_TLS_VOICE_PORT: '8444',
    });
    expect(env.ports.tls.default).toBe(8443);
    expect(env.ports.tls.get('voice')).toBe(8444);
  });

  it('returns undefined for missing or invalid values', () => {
    const env = readEnv({ GAMEFLOW_DEFAULT_PORT: 'not-a-port', GAMEFLOW_EMPTY_PORT: '' });
    expect(env.ports.default).toBeUndefined();
    expect(env.ports.get('empty')).toBeUndefined();
    expect(env.ports.get('missing')).toBeUndefined();
    expect(env.ports.tls.default).toBeUndefined();
  });

  it('exposes region and build id', () => {
    const env = readEnv({ GAMEFLOW_REGION: 'us-east-1', GAMEFLOW_BUILD_ID: 'build-42' });
    expect(env.region).toBe('us-east-1');
    expect(env.buildId).toBe('build-42');
  });
});

File: sdk/go/env_test.go
package gameflow

import "testing"

func TestNormalizePortName(t *testing.T) {
	cases := map[string]string{
		"voice chat": "VOICE_CHAT",
		"  default ": "DEFAULT",
		"Game":       "GAME",
	}
	for in, want := range cases {
		if got := normalizePortName(in); got != want {
			t.Errorf("normalizePortName(%q) = %q, want %q", in, got, want)
		}
	}
}

func TestParsePort(t *testing.T) {
	valid := map[string]uint16{"7777": 7777, " 8080 ": 8080}
	for in, want := range valid {
		got, ok := parsePort(in)
		if !ok || got != want {
			t.Errorf("parsePort(%q) = (%d, %v), want (%d, true)", in, got, ok, want)
		}
	}
	for _, in := range []string{"0", "-1", "abc", "99999", ""} {
		if got, ok := parsePort(in); ok {
			t.Errorf("parsePort(%q) = (%d, true), want ok=false", in, got)
		}
	}
}

func TestReadsNamedAndTLSPortsAndMetadata(t *testing.T) {
	env := mapEnv(map[string]string{
		"GAMEFLOW_DEFAULT_PORT":     "7777",
		"GAMEFLOW_VOICE_CHAT_PORT":  "7778",
		"GAMEFLOW_TLS_DEFAULT_PORT": "8443",
		"GAMEFLOW_REGION":           "eu-west-1",
		"GAMEFLOW_BUILD_ID":         "build-42",
	})
	ports := Ports{env: env}

	if got, ok := ports.Default(); !ok || got != 7777 {
		t.Errorf("Default() = (%d, %v), want (7777, true)", got, ok)
	}
	if got, ok := ports.Get("voice chat"); !ok || got != 7778 {
		t.Errorf("Get(voice chat) = (%d, %v), want (7778, true)", got, ok)
	}
	if _, ok := ports.Get("missing"); ok {
		t.Error("Get(missing) should be absent")
	}
	if got, ok := ports.TLS().Default(); !ok || got != 8443 {
		t.Errorf("TLS().Default() = (%d, %v), want (8443, true)", got, ok)
	}
	if got, _ := env.get("GAMEFLOW_REGION"); got != "eu-west-1" {
		t.Errorf("region = %q, want eu-west-1", got)
	}
	if got, _ := env.get("GAMEFLOW_BUILD_ID"); got != "build-42" {
		t.Errorf("buildID = %q, want build-42", got)
	}
}
